### PR TITLE
support HTTP Attack on chaosd

### PR DESCRIPTION
**Pull request by Andrewmatilde**
Signed-off-by: andrewmatilde <davis6813585853062@outlook.com>
Please install [chaos-tproxy](https://github.com/chaos-mesh/chaos-tproxy) before use the HTTP attack.
usage:
```
./bin/chaosd attack http abort -h
abort selected HTTP Package

Usage:
  chaosd attack http abort [flags]

Flags:
  -c, --code int32          Code is a rule to select target by http status code in response.
  -h, --help                help for abort
  -m, --method string       HTTP method
      --path string         Mathc path of Uri with wildcard matches.
      --port int32          port is a rule to select server listening on specific port.
  -p, --proxy_ports uints   composed with one of the port of HTTP connection, we will only attack HTTP connection with port inside proxy_ports (default [])
  -t, --target string       HTTP target: Request or Response

Global Flags:
      --log-level string   the log level of chaosd. The value can be 'debug', 'info', 'warn' and 'error'
      --uid string         the experiment ID
```
```
./bin/chaosd attack http delay -h
delay selected HTTP Package

Usage:
  chaosd attack http delay [flags]

Flags:
  -c, --code int32          Code is a rule to select target by http status code in response.
  -d, --delay time string   Delay represents the delay of the target request/response.
  -h, --help                help for delay
  -m, --method string       HTTP method
      --path string         Mathc path of Uri with wildcard matches.
      --port int32          port is a rule to select server listening on specific port.
  -p, --proxy_ports uints   composed with one of the port of HTTP connection, we will only attack HTTP connection with port inside proxy_ports (default [])
  -t, --target string       HTTP target: Request or Response

Global Flags:
      --log-level string   the log level of chaosd. The value can be 'debug', 'info', 'warn' and 'error'
      --uid string         the experiment ID
```
```
./bin/chaosd attack http file -h
attack with config file

Usage:
  chaosd attack http file [flags]

Flags:
  -p, --file path string   Config file path.
  -h, --help               help for file

Global Flags:
      --log-level string   the log level of chaosd. The value can be 'debug', 'info', 'warn' and 'error'
      --uid string         the experiment ID
```

### CODE REVIEW

**Comment by Andrewmatilde** [comment]
PTAL @WangXiangUSTC

File: pkg/server/server.go [hunk 1]
```diff
@@ -36,3 +42,12 @@ var Module = fx.Options(
 		scheduler.NewScheduler,
 	),
 )
+
+func provideNIl() (prometheus.Registerer, logr.Logger) {
```
Comment (WangXiangUSTC) [inline pkg/server/server.go:46]: what is the `NIl` means?

File: pkg/server/chaosd/http.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		return errors.Wrap(err, "create stdout pipe")
+	}
+
+	err = cmd.Start()
+	if err != nil {
+		return errors.Wrapf(err, "start command `%s`", cmd.String())
+	}
+
+	config, err := json.Marshal(&attackConf.Config)
+	fmt.Println(string(config))
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/http.go:61]: please use log instead of `fmt.Print`

**Comment by WangXiangUSTC** [comment]
@Hexilee PTAL

File: pkg/server/server.go (hunk 1 above)
Comment (Andrewmatilde) [reply pkg/server/server.go]: would I rename it to provide()?

File: pkg/server/chaosd/http.go (hunk 2 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go:61]: OK

File: pkg/server/server.go (hunk 1 above)
Comment (WangXiangUSTC) [reply pkg/server/server.go]: It is updated in previous pr, just ignore it here

File: pkg/server/chaosd/http.go (hunk 2 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go:61]: Fixed.

File: cmd/attack/http.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		"composed with one of the port of HTTP connection, "+
+			"we will only attack HTTP connection with port inside proxy_ports")
+	cmd.Flags().StringVarP((*string)(&o.Rule.Target), "target", "t", "",
+		"HTTP target: Request or Response")
+}
+
+func setSelector(cmd *cobra.Command, c *core.HTTPAttackOption) {
+	cmd.Flags().Int32Var(c.Rule.Selector.Port, "port", 0, "port is a rule to select server listening on specific port.")
+	cmd.Flags().StringVar(c.Rule.Selector.Path, "path", "",
+		"Mathc path of Uri with wildcard matches.")
```
Comment (WangXiangUSTC) [inline cmd/attack/http.go:99]: Is `Mathc` a typo?

File: cmd/attack/http.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+func setTarget(cmd *cobra.Command, o *core.HTTPAttackOption) {
+	cmd.Flags().UintSliceVarP(&o.ProxyPorts, "proxy_ports", "p", nil,
+		"composed with one of the port of HTTP connection, "+
+			"we will only attack HTTP connection with port inside proxy_ports")
+	cmd.Flags().StringVarP((*string)(&o.Rule.Target), "target", "t", "",
+		"HTTP target: Request or Response")
+}
+
+func setSelector(cmd *cobra.Command, c *core.HTTPAttackOption) {
+	cmd.Flags().Int32Var(c.Rule.Selector.Port, "port", 0, "port is a rule to select server listening on specific port.")
```
Comment (WangXiangUSTC) [inline cmd/attack/http.go:97]: how about refining the description?

File: pkg/server/chaosd/recover.go [hunk 5]
```diff
@@ -69,6 +69,8 @@ func (s *Server) RecoverAttack(uid string) error {
 			attackType = RedisAttack
 		case core.FileAttack:
 			attackType = FileAttack
+		case core.HTTPAttack:
+			attackType = AttackHTTP
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/recover.go:73]: how about changing to `HTTPAttack`?

File: pkg/core/http.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+func (c HTTPAttackConfig) RecoverData() string {
+	data, _ := json.Marshal(c)
+	return string(data)
+}
+
+type HTTPAttackOption struct {
+	CommonAttackConfig
+	ProxyPorts []uint                            `json:"proxy_ports,omitempty"`
+	Rule       tproxyconfig.PodHttpChaosBaseRule `json:"rules"`
```
Comment (WangXiangUSTC) [inline pkg/core/http.go:59]:
```suggestion
	Rule       tproxyconfig.PodHttpChaosBaseRule `json:"rule"`
```

File: pkg/core/http.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+const (
+	TargetRequest  tproxyconfig.PodHttpChaosTarget = "Request"
+	TargetResponse tproxyconfig.PodHttpChaosTarget = "Response"
+)
+
+const (
+	HTTPAbortAction = "abort"
+	HTTPDelayAction = "delay"
+	HTTPFileAction  = "file"
```
Comment (WangXiangUSTC) [inline pkg/core/http.go:38]: I think it's better to change `file` to `config`

File: pkg/server/chaosd/http.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+type attackHTTP struct{}
+
+var AttackHTTP AttackType = attackHTTP{}
+
+func (attackHTTP) Attack(options core.AttackConfig, _ Environment) error {
+	var attackConf *core.HTTPAttackConfig
+	var ok bool
+	if attackConf, ok = options.(*core.HTTPAttackConfig); !ok {
+		return errors.Errorf("AttackConfig -> *HTTPAttackConfig meet error")
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/http.go:41]: It's better to use `errors.New`

File: pkg/server/chaosd/http.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+	by, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Wrapf(err, "cannot read resp body")
+	}
+	fmt.Println(string(by))
+
+	attackConf.ProxyPID = cmd.Process.Pid
+	err = cmd.Process.Release()
+	if err != nil {
+		return errors.Wrapf(err, "Fatal error : release process fail , please clear PID: %d", attackConf.ProxyPID)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/http.go:97]:
```suggestion
		return errors.Wrapf(err, "Fatal error : release process fail, please clear PID: %d", attackConf.ProxyPID)
```

File: pkg/server/chaosd/http.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	by, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Wrapf(err, "cannot read resp body")
+	}
+	fmt.Println(string(by))
+
+	attackConf.ProxyPID = cmd.Process.Pid
+	err = cmd.Process.Release()
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/http.go:95]: why need to release the tproxy process?

File: cmd/attack/http.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+	setSelector(cmd, o)
+
+	delay := ""
+	o.Rule.Actions.Delay = &delay
+	cmd.Flags().StringVarP(o.Rule.Actions.Delay, "delay time", "d", "", "Delay represents the delay of the target request/response.")
+	return cmd
+}
+
+func setTarget(cmd *cobra.Command, o *core.HTTPAttackOption) {
+	cmd.Flags().UintSliceVarP(&o.ProxyPorts, "proxy_ports", "p", nil,
```
Comment (WangXiangUSTC) [inline cmd/attack/http.go:89]: how about changing to `proxy-ports` to keep the naming styles consistent

File: pkg/server/chaosd/http.go (hunk 2 above)
Comment (WangXiangUSTC) [reply pkg/server/chaosd/http.go:61]: Seems that it's not updated, and there are many of `fmt.Print` elsewhere, too

Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go:61]: Sorry  , I miss it.

File: cmd/attack/http.go (hunk 3 above)
Comment (Andrewmatilde) [reply cmd/attack/http.go:99]: Yep.

File: pkg/server/chaosd/http.go (hunk 2 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go]: Fixed.

File: cmd/attack/http.go (hunk 3 above)
Comment (Andrewmatilde) [reply cmd/attack/http.go]: Fixed.

File: cmd/attack/http.go (hunk 4 above)
Comment (Andrewmatilde) [reply cmd/attack/http.go:97]: How about `The TCP port that the target service listens on.` ? It is description in HTTP Chaos updated.

File: pkg/server/chaosd/recover.go (hunk 5 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/recover.go:73]: OK

File: pkg/core/http.go (hunk 7 above)
Comment (Andrewmatilde) [reply pkg/core/http.go:38]: OK

File: pkg/server/chaosd/http.go (hunk 8 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go:40]: OK

File: pkg/server/chaosd/http.go (hunk 9 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go:96]: OK

File: pkg/server/chaosd/http.go (hunk 10 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go:94]: Because chaos-tproxy in a proxy, we need to protect it from shutdown.

File: cmd/attack/http.go (hunk 11 above)
Comment (Andrewmatilde) [reply cmd/attack/http.go:89]: OK

File: pkg/server/chaosd/recover.go (hunk 5 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/recover.go]: Fixed.

File: pkg/core/http.go (hunk 6 above)
Comment (Andrewmatilde) [reply pkg/core/http.go]: Fixed.

File: pkg/core/http.go (hunk 7 above)
Comment (Andrewmatilde) [reply pkg/core/http.go]: Fixed.

File: pkg/server/chaosd/http.go (hunk 8 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go]: Fixed.

File: pkg/server/chaosd/http.go (hunk 9 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go]: Fixed.

File: cmd/attack/http.go (hunk 11 above)
Comment (Andrewmatilde) [reply cmd/attack/http.go]: Fixed.

File: cmd/attack/http.go (hunk 4 above)
Comment (WangXiangUSTC) [reply cmd/attack/http.go:97]: LGTM

Comment (Andrewmatilde) [reply cmd/attack/http.go]: Fixed.

File: pkg/server/chaosd/http.go (hunk 10 above)
Comment (Andrewmatilde) [reply pkg/server/chaosd/http.go:97]: Add a comment

**Review by WangXiangUSTC — APPROVED** [review]
LGTM

File: cmd/attack/http.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		NewHTTPConfigCommand(dep, option),
+	)
+
+	return cmd
+}
+
+func NewHTTPAbortCommand(dep fx.Option, o *core.HTTPAttackOption) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "abort",
+		Short: "abort selected HTTP Package",
```
Comment (Hexilee) [inline cmd/attack/http.go:55]: I think "Package" is not a normal concept in HTTP.

```suggestion
		Short: "abort selected HTTP message",
```
or
```suggestion
		Short: "abort selected HTTP connection",
```

File: pkg/core/http.go [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+func (c HTTPAttackConfig) RecoverData() string {
+	data, _ := json.Marshal(c)
+	return string(data)
+}
+
+type HTTPAttackOption struct {
+	CommonAttackConfig
+	ProxyPorts []uint                            `json:"proxy_ports,omitempty"`
+	Rule       tproxyconfig.PodHttpChaosBaseRule `json:"rule"`
```
Comment (Hexilee) [inline pkg/core/http.go:59]: How about using a new type to register options? I think we do not need the whole `PodHttpChaosBaseRule`, the several fields can meet our requirements. And more importantly, the nullable field is not suitable for command-line options.

File: pkg/core/http.go [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+func (o *HTTPAttackOption) PreProcess() (*HTTPAttackConfig, error) {
+	var c tproxyconfig.Config
+	zapLogger, err := zap.NewDevelopment()
+	if err != nil {
+		return nil, err
+	}
+	logger := zapr.NewLogger(zapLogger).WithName("HTTP Attack")
+	switch o.CommonAttackConfig.Action {
+	case HTTPAbortAction, HTTPDelayAction:
+		if *o.Rule.Selector.Path == "" {
```
Comment (Hexilee) [inline pkg/core/http.go:94]: As mentioned above, we need more complex code to deal with the `PodHttpChaosBaseRule`.

**Review by Hexilee — COMMENTED** [review]
Rest LGTM

File: pkg/core/http.go (hunk 14 above)
Comment (Andrewmatilde) [reply pkg/core/http.go:94]: I agree with you.

File: cmd/attack/http.go (hunk 12 above)
Comment (Andrewmatilde) [reply cmd/attack/http.go:55]: connection

File: pkg/core/http.go (hunk 13 above)
Comment (Andrewmatilde) [reply pkg/core/http.go:59]: I agree with you.

Comment (Andrewmatilde) [reply pkg/core/http.go]: Fixed.

File: pkg/core/http.go (hunk 14 above)
Comment (Andrewmatilde) [reply pkg/core/http.go]: Fixed.

**Review by Hexilee — APPROVED** [review]
LGTM

**Pull request merged**